### ISSUE TITLE
$push element to the 2nd level embedded array failed with 'TypeError: Cannot read property '$isMongooseDocumentArray' of undefined'

### ISSUE THREAD

**Issue by superZhen** [Automattic/mongoose#4645]
Dear all,

My mongoose is 4.6.4, and i have below schema:
```
var activitySchema = new mongoose.Schema({
    rounds              : {type : Number, required : true, min : 1},   //should be unique 
    name                : {type : String,  default : 'study activity'},
    totalTimes          : {type : Number,  default : 14, min : 1},
    description         : {type : String, required : true},
    startDate           : {type : Date,    default : Date.now()},     

    activities            : [
        {
            times       : {type : Number, required : true, min : 1},    //should be unique in 'activities'
            students    : [
                {
                    sid         : {type : mongoose.Schema.ObjectId},  //should be unique in 'students'
                    sname       : {type : String}, 
                    staffName   : {type : String},
                    lessonId    : {type : String}, 
                    ups         : {type : Number}, 
                    picUrl      : {type : String},
                    feedback    : {type : String},
                    status      : {type : Number, default: 0}
                }
            ]
        }
    ]
}, {timestamps: true})

schema.index({rounds: 1}, {unique : true});

```
i want to push element to 'activities' array and 'students' array separately, 
1. to add element for 'activities' array i need first check if 'times' existed or not, i did this in below 'else if branch', it works well;  
2. when related 'times' activity exists, i need push element to 'students' array,  i did this in below 'else branch', it failed with '**TypeError: Cannot read property '$isMongooseDocumentArray' of undefined**'.  i am not familiar with the nested array operation, and this have cost me 1 day. i  found this [article](http://stackoverflow.com/questions/32814436/mongodb-pushing-to-nested-array-use-0-not?noredirect=1&lq=1) is similar to mine, but still not work, thanks for anyone who can help me.
```
   var filter = {_id : aidObj, 'activities.times' : timesInt} //aidObj is the UUID of the whole document
   StudyActivity.findOne(filter, function (err, doc) { 
         if (err) {
                  console.log('addStudentToActivity: db findOne err = ' + err)
                  return res.json({code : 117, msg : 'db update error', err : err.errors})
          } else if (!doc) {
                   var activitity = {
                          times       : timesInt,
                          students    : [student]
                   }

                   filter = {_id : aidObj, 'activities.times' : {$ne : timesInt}}
                   update = {$push : {activities: activitity}}
                   option = {upsert : true, new :true}

                  StudyActivity.findOneAndUpdate(filter, update, option, function (err, activity) {
                            if (err) {
                                  console.log('addStudentToActivity: db update err = ' + err)
                                  return res.json({code : 117, msg : 'db ope failed', err : err})
                            } else {
                                  return res.json({code : 100, msg : 'add activity ok', activity: activity})
                            }
                   })
           } else {
                    filter = {_id : aidObj, 'activities' : {$elemMatch : {times : timesInt, 'students.sid' : {$ne : sidObj}}}}
                   var update = {$push : {'activities.$.students': student}}
                   var option = {upsert : true, new :true}

                    StudyActivity.update(filter, update, option, function (err, activity) {
                           if (err) {
                                 console.log('addStudentToActivity findOneAndUpdate err = ' + err)
                                 return res.json({code : 117, msg : 'db ope error', err : err})
                           } else {
                                 return res.json({code : 100, msg : 'add student succeed', activity: activity})
                           }
                     })
           }
                   
```

**Comment by superZhen** [comment]
some update:
if i change 
```
var update = {$push : {'activities.$.students': student}}          
```
to 
```
var update = {$push : {'activities.0.students': student}}
```
the update for 'students' will work.

**Comment by vkarpov15** [comment]
This error was fixed in 4.6.5, probably a duplicate of #4514. Upgrade and it should work :beers:

**Comment by haolucky** [comment]
thank you man 👍
